[#68225] use item formatter to display items

Kharonus opened the pull request:
# Ticket
[#68225](https://community.openproject.org/work_packages/68225)

# What are you trying to accomplish?
- show correctly formatting numbers in tree views and hierarchy admin forms

# What approach did you choose and why?
- use item formatter for tree views, item view components and standard to_s
- add deprecation warning for stringifying methods without new formatting

Kharonus committed:
[#68225] use item formatter to display items

- https://community.openproject.org/work_packages/68225
- use item formatter for tree views, item view components and standard to_s
- add deprecation warning for stringifying methods without new
  formatting

NobodysNightmare commented on app/components/admin/custom_fields/hierarchy/item_component.rb:
```diff
@@ -45,6 +45,15 @@ def wrapper_uniq_by
+            .new(label: false,
+                 number_length_limit: 42,
+                 number_integer_digit_limit: 40,
+                 number_precision: 40)
```
Wait what? Those numbers look utterly high. Do we really want to allow soooooooooooooooooooooooooooooooooooooo long numbers? Especially since I remember that we do not make use of thousands separators, so the numbers will look like 200000000000000000000000000000000000000 or 0.0000000000000000000000000000000000001.

At this point I would even need to check the precision limits of a 64 bit floating point number... As far as I can tell integers can't be precisely represented in this scale at all, I am not yet sure about the very small numbers, but let me say that I would be surprised if they were represented with full precision.

Kharonus replied:
those are `BigDecimal`, not `float64`. And yes, there are precision issue. The last requirement I got from UX was, that in this component the numbers should have a "huge" limit before converting into scientific notation. But I can reiterate on that requirement.

NobodysNightmare replied:
Thanks for reminding me, I was in float-land with my thoughts.

Opinions from my side:

* these numbers require thousands separators (naively assuming that big numbers are more common than equally small numbers)
* Nobody thought about rendering for numbers requiring this amount of horizontal space

But it's okay, I'll not fight about this, because if anyone ever inputs so large numbers the fight will resolve itself :D

NobodysNightmare commented on app/models/custom_field/hierarchy/item.rb:
```diff
@@ -46,9 +47,16 @@ def ancestry_path(include_shorts_and_weights: false)
+  private
+
+  def formatter
+    @formatter ||= CustomFields::Hierarchy::HierarchicalItemFormatter.new
```
:yellow_circle: This would cause the instantiation of a comparatively large object for every call to `to_s` on the item class (per item). Since this is a completely vanilla/default formatter, what do you think about making this available as `CustomFields::Hierarchy::HierarchicalItemFormatter.default`?

Kharonus replied:
you mean, `default` should be a class method? so that we do not instantiate this object?

to be very correct, this object would be only instantiated once per item, as it is memoized on it. Only reinstantiate the item would lead to reinstantiating the formatter.

Another option would be to not use the formatter in `to_s`. But I thought it would be a nice idea. WDYT?

NobodysNightmare replied:
> you mean, default should be a class method?

Yes, that's how I intended it. Essentially:

```ruby
class << self
  def default
    @default ||= new
  end
end
```

> to be very correct, this object would be only instantiated once per item, as it is memoized on it

This is correct. But for code like `items.each { |item| puts item.to_s }` we would create many formatters.

> Another option would be to not use the formatter in to_s. But I thought it would be a nice idea. WDYT?

I am open to both ways. You can argue against the item knowing its formatter, because the item shouldn't care about that at all. On the other hand you can argue that we shouldn't duplicate code to format the item as a string, so of course the item should use it.

Kharonus replied:
done

NobodysNightmare commented on app/services/custom_fields/hierarchy/hierarchical_item_service.rb:
```diff
@@ -106,6 +106,14 @@ def get_branch(item:)
         Success(item.self_and_ancestors.reverse)
       end
 
+      # Gets all nodes in a tree from the item/node back to the root without the item/node itself.
```
:red_circle: 

<img width="419" height="238" alt="image" src="https://github.com/user-attachments/assets/4bc70110-a5d9-4c94-8976-e0a937183fdd" />

The implementation of this method is exactly the same as for `get_branch`, so I'd assume it to return exactly the same result (nodes **including** the item itself).

NobodysNightmare replied:
(I also just noticed that there is no spec for the path-only case)

Kharonus replied:
Ah, crap, I forgot to amend that :D I had it planned both, changing the closure tree method AND writing a test case.

Kharonus replied:
done

NobodysNightmare commented on app/services/custom_fields/hierarchy/hierarchical_item_formatter.rb:
```diff
@@ -0,0 +1,105 @@
+      def format(item:)
+        parts = []
+
+        if @path && @label
```
~~:yellow_circle: Why is this constructed as a three-case if-elsif-elsif? As far as I can tell, we only use two inputs that may or may not be concatenated.~~

I'd have expected something like

```ruby
parts << ancestors(item) if @path
parts << item.label if @label
```

edit: Okay I see it now. `parts` is for all visible parts of the formatted thing, not just the path components and concatenating the path with `/` already happens in `branch` and `ancestors`... I guess it makes sense, but it surprised me.

Kharonus replied:
I stumbled across this, too. But I didn't want to concatenate again after already concatenating the path. 

Another Idea would be to actually get a list of items back (only ancestors or ancestors plus label) and then doing the concatenation here.

I think, I like that. Let me try, how it looks in code.

Kharonus replied:
I changed a bit ;)

NobodysNightmare commented on app/services/custom_fields/hierarchy/hierarchical_item_formatter.rb:
```diff
@@ -0,0 +1,105 @@
+      def ancestors(item) = persistence_service.get_ancestors(item:).value!.filter_map(&:label).join(" / ")
+
+      def format_suffix(item)
+        return "" if item.short.nil? && item.weight.nil?
```
Will short ever be nil for a hierarchy? Since users submit items through a form that has an input for the short, I'd expect them to submit an empty string and thus I'd expect short to always be a string that sometimes is `""`.

In other words: I'd probably rewrite this as

```suggestion
        return nil if item.short.blank? && item.weight.nil?
```

Kharonus replied:
So, in general, we should NEVER store an empty string to a short. That's what our contracts should guard us for. A short is either present or nil. And it is mutual exclusive with a weight, that is either a present or nil.

The only difference in the contracts is, that the list item is valid without a short (nil), while a weighted list item must have a weight.

NobodysNightmare replied:
Indeed. I didn't expect this, but you are right. Then let's keep it as strict as you wrote it.

NobodysNightmare commented on spec/services/custom_fields/hierarchy/hierarchical_item_formatter_spec.rb:
```diff
@@ -0,0 +1,86 @@
+    end
+  end
+
+  context "with specific number formatting" do
```
I'd probably make this context **only** about the special number formatting and not _also_ about the parenthesis, but rather create a separate context for that, which also tests the missing parenthesis on shorts.

Kharonus replied:
true

Kharonus replied:
done

Kharonus committed:
[#68225] reworked format for hierarchy items

- introduce default formatter instance
- added missing specs for persistence service methods

NobodysNightmare commented on app/services/custom_fields/hierarchy/hierarchical_item_formatter.rb:
```diff
@@ -0,0 +1,106 @@
+        path << item.label if @label
+
+        str_parts = []
+        str_parts << path.join(" / ") unless path.empty?
```
:green_circle: I'd have a preference for fewer negations. I actually thought briefly that the condition is the wrong way around, but it isn't of course ^^

```suggestion
        str_parts << path.join(" / ") if path.present?
```

(technically `present?` checks more things than `!empty?`, because it's defined on many types; but if you know that something is an array, it's effectively just that)

NobodysNightmare reviewed (approved):
I had one minor remark on the newly edited code, but this LGTM now :)

Kharonus committed:
Merge branch 'dev' into implementation/68225-represent-scores-in-tree-view-components-and-project-attribute-sections

Kharonus committed:
[#68225] fixed test expectation
